include/zephyr/drivers/fuel_gauge.h

```diff
@@ -156,9 +170,22 @@ struct fuel_gauge_get_property {
 		uint16_t sbs_at_rate_time_to_empty;
 		/** FUEL_GAUGE_SBS_ATRATE_OK */
 		bool sbs_at_rate_ok;
+		/** FUEL_GAUGE_MANUFACTURE_DATE */
+		uint16_t manufacturer_date;
+		/** FUEL_GAUGE_SERIAL_NUMBER */
+		uint16_t serial_number;
 	} value;
 };
 
+/** Block properties are separated due to size */
+struct fuel_gauge_get_block_property {
+	/** Battery fuel gauge property to get */
+	uint16_t property_type;
+
```

Comment:
The fuel gauge doesn't return a null terminated string - it returns a block of bytes, with the first character being the actual length, and the rest of the bytes being the message.  So, to create a null terminated string, the application would need to do something like:
		sprops.property_type = FUEL_GAUGE_MANUFACTURER_NAME;
		ret = fuel_gauge_get_block_prop(dev, &sprops, SBS_GAUGE_MANUFACTURER_NAME_MAX_SIZE, buffer);
		if (ret < 0) {
			printk("Error: cannot get properties\n");
		} else {
			if (ret != 0) {
				printk("Warning: get block property failed\n");
			}

			if (sprops.status == 0) {
				**buffer[(int)buffer[0]+1] = 0;**
				printk("Mfg name %s\n", &buffer[1]);
			}
		}

That said, I was thinking of actually adding a length field, not specifically to tell the user the actual length, but rather to limit/protect the user.  For instance, if the user requests a read of 1000 bytes, compare against the property type, and take the MIN(userlen, propertyTypeMAXLENGTH) as the return value.

I guess another check the driver could do is to look at that first byte, and if it's greater than the above min - say the user only gives a buf_len of 3, but the first byte is 10, should probably return a -EFAULT error or something.

Then the question would be - should the driver read the first byte and set return length, and then shift the byte array, so that the length is no longer encoded in the return?  On the one hand, it makes it easier for the user.  On the other hand, the user is probably expecting the raw string back from the fuel gauge, so the driver is kind of over-stepping it's boundaries.

Reply:
@protterc 

Taking a step back, we might want to solve this by bringing the union back.

What if we go back to using a union with array pointers of different types that get defined by our API? So a use case would look something like:

```
fuel_gauge_get_block_prop(dev, &sprops, SBS_GAUGE_MANUFACTURER_NAME_MAX_SIZE, buffer);

struct abstract_type *foo = sprops[0].value.field.
```

Here the driver would just set sprops[0].value to the buffer address. This would allow the API to abstract away the data as something the client can more easily reason with and would be less IC-specific.

So here we'd be letting the client allocate the memory for us but we get to use our IC-specific knowledge on how to get it and then just give a presentation to the client. It also increases the similarity between the two APIs.

Let me know if you want me to provide a more fully fleshed out example.

Reply:
Maybe a more complete explanation would help.  Not sure I understand quite what you're thinking.

Reply:
@protterc 

Here's a more complete example of what I was thinking. I'd like if drivers can expose arbitrary data to the client and present it as whatever type we feel is abstract and appropriate for them.

```c
/*****************************************************************************/
/*                  Goal: Expose block properties to client                  */
/*****************************************************************************/

#define FUEL_GAUGE_SB_MANUF_NAME 0x1234
#define FUEL_GAUGE_SOME_PROP	 0x1235

/* We could just expose this as a NULL terminated string too */
struct sb_string {
	/** Number of characters in contents */
	uint8_t size;
	/** Contents of sb string minus the size byte */
	char contents[31];
};

struct some_other_type {
	uint32_t foo;
	uint16_t bar;
	uint16_t zeeb;
};

struct fuel_gauge_get_block_property {
	/** Battery fuel gauge property to get */
	uint16_t property_type;
	/** Negative error status set by callee e.g. -ENOTSUP for an unsupported property */
	int status;
	/** Pointer to address in user allocated memory; value is set by driver */
	union {
		/** Raw pointer that is useful for custom types */
		void *raw;
		struct sb_string *manufacturer_name;
		struct some_other_type *some_property;
	} value;
};

/*
 * Function prototype that allows multi-property fetch with one call. We don't need to do this but
 * we can.
 */
int fuel_gauge_get_block_prop(const struct device *dev, struct fuel_gauge_get_block_property *props,
			      size_t props_len, size_t dst_len, void *dst);

/* API Defined above with example usage below */
ZTEST_USER_F(sbs_gauge_new_api, test_block_prop_api_demo)
{

	/* Client use-case with anonumous struct *****************************/
	struct {
		struct sb_string my_manufacturer_name;
		struct some_other_type my_some_property;
	} my_prop_buffer;

	struct fuel_gauge_get_block_property props[] = {
		{
			.property_type = FUEL_GAUGE_SB_MANUF_NAME,
		},
		{
			.property_type = FUEL_GAUGE_SOME_PROP,
		},
	};

	fuel_gauge_get_block_prop(fixture->dev, props, ARRAY_SIZE(props), sizeof(my_prop_buffer),
				  &my_prop_buffer);

	/* These have to be address equal */
	zassert_equal_ptr(props[0].value.manufacturer_name, &my_prop_buffer.my_manufacturer_name);
	zassert_equal_ptr(props[1].value.some_property, &my_prop_buffer.my_some_property);

	/* Client use-case with allocated buffer *****************************/
	uint8_t some_max_bytes = 64;
	uint8_t buffer[some_max_bytes];

	struct fuel_gauge_get_block_property props2[] = {
		{
			.property_type = FUEL_GAUGE_SB_MANUF_NAME,
		},
		{
			.property_type = FUEL_GAUGE_SOME_PROP,
		},
	};

	fuel_gauge_get_block_prop(fixture->dev, props2, ARRAY_SIZE(props2), sizeof(buffer), buffer);

	struct sb_string *my_manuf_name2 = props2[0].value.manufacturer_name;
	struct some_other_type *my_some_property2 = props2[1].value.some_property;

	/* These have to be address equal */
	zassert_equal_ptr(my_manuf_name2, buffer);
	zassert_equal_ptr(my_some_property2, +sizeof(*my_manuf_name2));

	/*
	 * Note above that by using the pointer from the API, the client doesn't have to keep the
	 * properties and anonymous struct fields in order
	 */
}
```

Reply:
I'm have a few concerns around this interface.

To clarify my understanding, are the structs sb_string and some_other_type defined in the fuel_gauge.h file?  If so, I'm less worried about size, but if these are user-defined structs, then we could still have potential for memory corruption.

If they are created in fuel_gauge.h, I'm mixed.  On the one hand, they do a great job of defining what's coming back, but on the other hand, they really belong to sbs_gauge.h - they're not generic, but very specific to a given implementation (SBS), so having them in the general interface seems broken.

Having an array of props, but only a single buffer seems problematic in the sense that you need to make sure you order the elements in my_prop_buffer in the same order as the props array.  Also concerned that if there are size problems - like you forget to add a struct, but request a property - do we return an error for the whole batch of properties.  The first few might be OK, because they fit into the mixed buffer, but perhaps are not in the right memory location, depending on where the needed struct wasn't included.

Additionally, you've re-created, in a different way, the issue with consuming potentially a large amount of stack space required of you request all the properties at once.

I guess, given that these are generally static values, and only need to be retrieved once, and cached in the MCU firmware, are we overthinking, or should we just have a general block read function that takes a register, a buffer, and a size, and leave the details of how the data is mapped to the end user, which if they're using these drivers, probably knows what they need, and what they're going to get returned to them based on the data sheet.

Reply:
@protterc 

> Let's keep discussing, I'm sure we'll come up with the right interface in the end :-)

Of course! I appreciate the discussion and the time you've taken here along with
the quick responses. Thank you!


> To clarify my understanding, are the structs sb_string and some_other_type defined in the fuel_gauge.h file? 

Yes these were meant to be types defined in `fuel_gauge.h` they were sort of just meant to be rough examples of what could be exposed to a client.

> On the one hand, they do a great job of defining what's coming back, but on the other hand, they really belong to sbs_gauge.h - they're not generic,

Right, I probably should've made it less SBS-specific, but wanted to give a
spec-specific example and a generic example. I think we ought to keep things
generic and not specific to SBS or a particular IC when we reasonably can. We
pay no extra memory cost for adding extra/various struct definitions but we do
pay the maintenance cost of course.

> Having an array of props ...

You're right. Let's keep it to a single prop, it just reduces the complexity for
now and makes it easier for driver developers. We can always expand the API when
there's sufficient motivation to do so.

> Additionally, you've re-created, in a different way, the issue with consuming potentially a large amount of stack space required of you request all the properties at once.

That is correct if the client fetches all the properties at once, but then
that's really just on the client anyway. Although a client could just allocate
that memory in the .data section instead of the stack. **Regardless**, I think
you're right and that we ought to simplify the API by keeping the block
properties to single fetches for now.

> I guess, given that these are generally static values, and only need to be retrieved once

This is definitely true for properties like `manufacturer name` but might not be
true for other arbitrary data in manufacturer_data or some other IC that isn't
SBS.

> which if they're using these drivers, probably knows what they need

It's my intention to keep as much of the datasheet specifics to the driver
implementation and provide some kind of abstraction to the client. Of course, we
may have some leaky abstraction such as properties with an SBS prefix because
they're specific to SBS. But I think we should do best-effort to enable usage of
the API without familiarity with the underlying IC datasheets.


So how would you feel if we basically just dropped the whole multiple property part?

So something like:

```c
int fuel_gauge_get_block_prop(const struct device *dev, struct
fuel_gauge_get_block_property *props, size_t dst_len, void *dst);
```

Reply:
@aaronemassey - I think one property at a time makes sense for the block reads, and the API you suggest should work for any case.  

Adding a sbs specific area to the fuel_gauge.h file seems reasonable.  Basically a place to define all the block structures and sizes seems legit.  It stretches the architecture a little bit, but I think it's probably ok.  If people want to use them, great, if people just want to send a char buffer, that's good too.

Want to get your thoughts on one last thing.  RIght now, we support, by name, just reading the 3 string registers.  Given that the API will only read the size that's passed in, would it be OK to allow other property_type's to be sent to the block read function?  What I'm thinking is that I could send a register at the start of all the dynamic data I want to read, and a size, and get it all in one shot.  Would make it so that I can use block read, instead of having to add the logic to sort out the props array, determine if they're all sequential, and then do a block read.  It all seems safe enough, given the user has to provide the buffer and buffer size, just wanted to run it past you.

Peter

Reply:
@protterc

> If people want to use them, great, if people just want to send a char buffer,
> that's good too.

We can always enable something like a raw version of a property too. But I'd
rather keep the IC/implementation specifics away from the client and just within
the driver.

> What I'm thinking is that I could send a register at the start of all the
> dynamic data I want to read, and a size, and get it all in one shot...

So I'd like to stay away from clients having to give a specific register.

> Would make it so that I can use block read, instead of having to add the logic
> to sort out the props array, determine if they're all sequential, and then do
> a block read.

I think we should keep that to the current API as-is. Figuring out what
properties are contiguous or not can be done with nominal extra compute time in
the driver. It'll save considerable more time compared to the bus reads,
especially if it's SMBus.

I spent a bit of time thinking up of a rough idea on how to do this and I think
we can do the contiguous bus reads on the fly and even with more than a single
sequence of contiguous registers in a single function call. Some of this logic
can be in a shared fuel gauge library used by the drivers too. But this probably
it's own discussion :)